### PR TITLE
DTLS1.3/TLS1.3: Fix max_fragment_length extension parsing when using PSK

### PR DESCRIPTION
**Pull request by fwh-dc**
The issue: https://github.com/openssl/openssl/issues/18121 appeared in the DTLS test cases when enabling DTLS 1.3.

I've cherry-picked relevant bits from https://github.com/openssl/openssl/pull/18130.

This PR is dependent on https://github.com/openssl/openssl/pull/24426 for a binder calculation fix.

##### Checklist
- [x] tests are added or updated

### CODE REVIEW

**Comment by fwh-dc** [comment]
@t8m @mattcaswell This is ready for review.

**Comment by vdukhovni** [comment]
From a recent discussion on the IETF TLSWG list, it appears that we're much too strict in our processing of the (now deprecated) maximum fragment length extension. What we should do is:

* If we're resuming a session, *ignore* all RFC6066 extensions other than perhaps SNI, using the values from the resumed session. Any conflicting or missing values in the client HELLO don't matter.
* If we're doing a full handshake, we use the data from the client HELLO, and even if there were a candidate session we chose to not resume, the extensions in effect at that point don't matter.

So unless I'm misreading the context/content of this PR, it seems to me that instead of looking to make resumption detection more precise, and then reject differences in extensions in fewer cases, we should NEVER reject differences.

File: ssl/statem/extensions_srvr.c
Comment (nhorman) [inline ssl/statem/extensions_srvr.c:199]: Speaking very naively here, but, gien that max_fragment_len_mode gets set based on the parsing of extensions, doesn't control flow through this function get altered based on the order in which those extensions are processed? Is that the intent, or is it irrelevant?

Comment (vdukhovni) [reply ssl/statem/extensions_srvr.c:199]: So long as any session ticket extension is processed first, the rest don't matter. In any case, I do think that once we stop being needlessly (erroneously) pedantic, the logic changes a lot, because we either use the new value if not resuming, or keep the old value otherwise, and there's nothing to ever reject (based on this extension and values or not from a previous session).

**Comment by vdukhovni** [comment]
> From a recent discussion on the IETF TLSWG list, it appears that we're much too strict in our processing of the (now deprecated) maximum fragment length extension. What we should do is:

Link to the TLSWG thread: https://mailarchive.ietf.org/arch/msg/tls/IDmCa7UcoKpMyxG59FKbGG5g5C0/

**Comment by fwh-dc** [comment]
@vdukhovni 
> So unless I'm misreading the context/content of this PR, it seems to me that instead of looking to make resumption detection more precise, and then reject differences in extensions in fewer cases, we should NEVER reject differences.

Hmm. I'm not sure I understand you correctly so I want to add some context.

The issue that this PR is trying to fix is that the MaxFragmentLen extension does not work when using TLS1.3 PSK ciphersuites. The reason for this is that `s->hit` is set not only for session resumptions but also for psk sessions.

My understanding is that PSK is not the same as session resumption.
So instead of relying on checking `s->hit` instead we initialize the `s->session->ext.max_fragment_len_mode` to unspecified and set it to the previous sessions value if we are resuming.

I'm not sure if you propose that this logic needs updating?

@nhorman 
> Speaking very naively here, but, gien that max_fragment_len_mode gets set based on the parsing of extensions, doesn't control flow through this function get altered based on the order in which those extensions are processed? Is that the intent, or is it irrelevant?

That's a good point. It made me aware that the check 
```
    if (org_mfl_mode != s->session->ext.max_fragment_len_mode
            && org_mfl_mode != TLSEXT_max_fragment_length_UNSPECIFIED) {
        if (s->session->ext.max_fragment_len_mode == TLSEXT_max_fragment_length_UNSPECIFIED) {
            s->session->ext.max_fragment_len_mode = org_mfl_mode;
        } else {
            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER,
                     SSL_R_SSL3_EXT_INVALID_MAX_FRAGMENT_LENGTH);
            goto err;
        }
    }
```

Was performed before extension parsing. I've made a fixup for that part.

**Comment by vdukhovni** [comment]
My point is that we also seem to be *rejecting* new client hello values for the extension, when a different value exists from a resumed session. Perhaps that's not logic you're intending to change, but it strongly overlaps with this PR. I'd like to see those checks gone. At which point your PR should be very different, because we accept the new value if not resuming (including a non-resumption PSK) and ignore it when resuming.

**Comment by fwh-dc** [comment]
> ..., because we accept the new value if not resuming (including a non-resumption PSK) and ignore it when resuming.

I get it now! I'll make a fix for this.

**Comment by fwh-dc** [comment]
@vdukhovni Please let me know if this is in line with your thoughts :)

**Comment by vdukhovni** [comment]
[ WIth some notable help from @mattcaswell, I believe I've arrived at a better understanding of the extension processing control flow. So with that I'd like to propose some correction and simplification ]

* There is no need to look at renegotiation, this either performs a resumption (abbreviated renegotiate) or a full handshake, and the fact that it happened as part of renegotiation is not material, all that matters is whether we end up with a new session or not.
* More importantly, the code just below your proposed (but will not be needed):
    ```
    if (s->session != NULL)
        org_mfl_mode = s->session->ext.max_fragment_len_mode;
    ```
  is where any session resumption takes place, and all further extension processing will take place in the context of the resumed session. That session will have MFL != UNSPECIFIED if resumed, and == UNSPECIFIED if new.  So all that needs to happen is that the MFL is used if and only if still == UNSPECIFIED. Then in final (as in your PR) it is set to DISABLED if still unspecified.

The PR then becomes simpler still!

**Comment by fwh-dc** [comment]
@vdukhovni 

Just to be sure:
So effectively what you are saying is:
If `s->session->ext.max_fragment_len_mode == TLSEXT_max_fragment_length_UNSPECIFIED` in `tls_parse_ctos_maxfragmentlen()` then we are NOT doing resumption. Else we are doing a resumption?

I've pushed a fix which assumes the above.

**Review by vdukhovni — APPROVED** [review]
LGTM. Yes, this now looks correct, fixing not only the misuse of the overloaded `s->hit`, but also improving RFC6066 conformance, and now much simpler. Thanks!

**Comment by vdukhovni** [comment]
At this point, please squash the fixups and rebase. The changes to the SCTP assertion comment are perhaps unrelated to this PR, and could be dropped. With that, we should be ready to merge.

File: test/dtls_mtu_test.c
Comment (fwh-dc) [reply test/dtls_mtu_test.c:232]: When sctp is enabled `ossl_statem_export_allowed()` gets called through `SSL_export_keying_material()` -> `tls13_export_keying_material()` when sending/receiving ServerHello and the ClientKeyExchange.

As Matt notes in https://github.com/openssl/openssl/pull/24426#issuecomment-2137382145 SCTP support in DTLSv1.3 is TBD. So instead of disabling the failing test completely I set the maximum version to DTLS1.2 to test for regressions.

Comment (fwh-dc) [reply test/dtls_mtu_test.c:232]: But maybe we should wait for https://github.com/openssl/openssl/pull/24605 to be merged and then I can rebase on top of that to revert the change completely.

**Comment by fwh-dc** [comment]
> At this point, please squash the fixups and rebase.

It was my understanding that your merge process did `--autosquash`?

**Comment by FdaSilvaYY** [comment]
This fix should better be push to master branch, IMHO ?

**Comment by fwh-dc** [comment]
> This fix should better be push to master branch, IMHO ?

I'd like the `feature/dtls-1.3` branch to be rebased after merging it to master in that case

**Comment by vdukhovni** [comment]
Once this PR has the necessary approvals, we can merge it into master, then rebasing the feature branch makes sense, and is perhaps good to do periodically anyway, to limit divergence. Not sure who does that, @mattcaswell may know the process.

**Comment by vdukhovni** [comment]
> > At this point, please squash the fixups and rebase.
> 
> It was my understanding that your merge process did `--autosquash`?

Any comment on the SCTP comment tweak? Is it intended to be part of this PR, or did it sneak in by accident?

**Comment by fwh-dc** [comment]
> Any comment on the SCTP comment tweak? Is it intended to be part of this PR, or did it sneak in by accident?

It is intentional. The CI fails (on `feature/dtls-1.3` branch) if it is left out.
See https://github.com/openssl/openssl/pull/24513#discussion_r1645379193

**Comment by vdukhovni** [comment]
> > Any comment on the SCTP comment tweak? Is it intended to be part of this PR, or did it sneak in by accident?
> 
> It is intentional. The CI fails (on `feature/dtls-1.3` branch) if it is left out. See [#24513 (comment)](https://github.com/openssl/openssl/pull/24513#discussion_r1645379193)

Ok then, justing waiting for final approvals, merge to master and rebase...

**Comment by mattcaswell** [comment]
Please drop the SCTP change (and move to a different PR) - and I will approve this.

> Once this PR has the necessary approvals, we can merge it into master, then rebasing the feature branch makes sense, and is perhaps good to do periodically anyway, to limit divergence. Not sure who does that, @mattcaswell may know the process.

I can do that. Someone please ping me once this is merged to remind me!

**Comment by fwh-dc** [comment]
@mattcaswell @vdukhovni Updated to target master

**Comment by t8m** [comment]
Is this something that should be considered as a bug fix and eligible for backport to stable branches?

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
> Is this something that should be considered as a bug fix and eligible for backport to stable branches?

It fixes #18121 - so yes.

**Review by t8m — APPROVED** [review]

**Comment by vdukhovni** [comment]
So this needs to be rebased onto master, rather than the dtls feature branch. Who will do that?

**Comment by t8m** [comment]
Reworded the commit message and merged to all the active branches. Thank you for your contribution.

**Comment by fwh-dc** [comment]
> I can do that. Someone please ping me once this is merged to remind me!

@mattcaswell merged now please remember to rebase `feature/dtls-1.3`

**Pull request closed without merging**
